### ISSUE TITLE
Returning from [[noreturn]] function not rejected in constant evaluation

### ISSUE THREAD

**Issue by GKxxUCAS** [llvm/llvm-project#211272]
Consider the following code:

```cpp
[[noreturn]] constexpr void f(int i) {
    if (i == 0)
        throw 42;
    // otherwise it returns, which is UB
}

constexpr int g() {
    f(1);
    return 0; // unreachable
}

int main() {
    static_assert(g() == 0, "This should cause a compile error due to UB");
    (void)g(); // avoid warning about unused function
    return 0;
}
```

From C++20 final draft N4861 [dcl.attr.noreturn]/2,
> If a function f is called where f was previously declared with the noreturn attribute and f eventually returns, the behavior is undefined.

So this code involves an undefined behavior during constant evaluation and should be rejected.

Neither GCC nor Clang rejects this. [godbolt](https://godbolt.org/z/Gjj5hv4qr)

**Comment by zwuis** [comment]
After [CWG2763](https://wg21.link/cwg2763), it is unspecified whether `g()` is a core constant expression. [CWG2924](https://wg21.link/cwg2924) changes 'unspecified' to 'implementation-defined'.
